extensions/catalog/federated-catalog-cache/src/test/java/org/eclipse/dataspaceconnector/catalog/cache/crawler/CrawlerImplTest.java

```diff
@@ -0,0 +1,214 @@
+package org.eclipse.dataspaceconnector.catalog.cache.crawler;
+
+import net.jodah.failsafe.RetryPolicy;
+import org.eclipse.dataspaceconnector.catalog.cache.TestProtocolAdapterRegistry;
+import org.eclipse.dataspaceconnector.catalog.cache.TestWorkItem;
+import org.eclipse.dataspaceconnector.catalog.cache.TestWorkQueue;
+import org.eclipse.dataspaceconnector.catalog.spi.CatalogQueryAdapter;
+import org.eclipse.dataspaceconnector.catalog.spi.CrawlerErrorHandler;
+import org.eclipse.dataspaceconnector.catalog.spi.WorkItem;
+import org.eclipse.dataspaceconnector.catalog.spi.WorkItemQueue;
+import org.eclipse.dataspaceconnector.catalog.spi.model.UpdateRequest;
+import org.eclipse.dataspaceconnector.catalog.spi.model.UpdateResponse;
+import org.eclipse.dataspaceconnector.spi.EdcException;
+import org.eclipse.dataspaceconnector.spi.monitor.Monitor;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+
+import java.time.Duration;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.easymock.EasyMock.anyString;
+import static org.easymock.EasyMock.expect;
+import static org.easymock.EasyMock.expectLastCall;
+import static org.easymock.EasyMock.isA;
+import static org.easymock.EasyMock.mock;
+import static org.easymock.EasyMock.niceMock;
+import static org.easymock.EasyMock.replay;
+import static org.easymock.EasyMock.strictMock;
+import static org.easymock.EasyMock.verify;
+
+class CrawlerImplTest {
+    public static final int QUEUE_CAPACITY = 3;
+    public static final int JOIN_WAIT_TIME = 100;
+    public static final int WORK_QUEUE_POLL_TIMEOUT = 500;
+    private CrawlerImpl crawler;
+    private CatalogQueryAdapter protocolAdapterMock;
+    private ArrayBlockingQueue<UpdateResponse> queue;
+    private Monitor monitorMock;
+    private WorkItemQueue workQueue;
+    private TestProtocolAdapterRegistry registry;
+    private CrawlerErrorHandler errorHandlerMock;
+    private ExecutorService executorService;
+
+    @BeforeEach
+    void setUp() {
+        executorService = Executors.newSingleThreadExecutor();
+        errorHandlerMock = mock(CrawlerErrorHandler.class);
+        protocolAdapterMock = strictMock(CatalogQueryAdapter.class);
+        queue = new ArrayBlockingQueue<>(QUEUE_CAPACITY);
+        monitorMock = niceMock(Monitor.class);
+        workQueue = new TestWorkQueue(10);
+        registry = new TestProtocolAdapterRegistry(protocolAdapterMock);
```

Comment:
In this case I would use a mock for the `CatalogQueryAdapterRegistry`, it will permit to avoid the re-instantiation of the crawler on the last test of the suite, and give more expressiveness to the test through mock behavior configuration